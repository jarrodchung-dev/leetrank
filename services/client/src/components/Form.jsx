import React from "react";
import { Redirect } from "react-router-dom";

const Form = (props) => {
  if (props.isAuthenticated) {
    return <Redirect to="/" />;
  }
  return (
    <div>
      {props.type === "Login" && <h1 className="title">Log In</h1>}
      {props.type === "Register" && <h1 className="title">Register</h1>}
      <form onSubmit={(event) => props.handleFormSubmit(event)}>
        {props.type === "Register" && (
          <div className="field">
            <label className="label">Username</label>
            <div className="control">
              <input
                name="username"
                value={props.data.username}
                className="input"
                type="text"
                onChange={props.handleChange}
              />
            </div>
          </div>
        )}
        <div className="field">
          <label className="label">Email</label>
          <div className="control">
            <input
              name="email"
              value={props.data.email}
              className="input"
              type="email"
              onChange={props.handleChange}
            />
          </div>
        </div>
        <div className="field">
          <label className="label">Password</label>
          <div className="control">
            <input
              name="password"
              value={props.data.password}
              className="input"
              type="password"
              onChange={props.handleChange}
            />
          </div>
        </div>
        <button className="button" type="submit">
          Submit
        </button>
      </form>
    </div>
  );
};

export default Form;
