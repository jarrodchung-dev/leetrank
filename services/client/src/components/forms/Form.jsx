import React, { Component } from "react";
import axios from "axios";
import { Redirect } from "react-router-dom";
import { registerFormRules, loginFormRules } from "./form-rules";
import FormErrors from "./FormErrors.jsx";

class Form extends Component {
  constructor(props) {
    super(props);
    this.state = {
      formData: {
        username: "",
        email: "",
        password: ""
      },
      registerFormRules: registerFormRules,
      loginFormRules: loginFormRules,
      valid: false
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.clearForm = this.clearForm.bind(this);
    this.allValid = this.allValid.bind(this);
    this.resetRules = this.resetRules.bind(this);
  }
  componentDidMount() {
    this.clearForm();
    this.validateForm();
  }
  componentDidUpdate() {}
  clearForm() {
    this.setState({
      formData: {
        username: "",
        email: "",
        password: ""
      }
    });
  }
  handleChange(event) {
    const formData = this.state.formData;
    formData[event.target.name] = event.target.value;
    this.setState(formData);
    this.validateForm();
  }
  handleSubmit(event) {
    event.preventDefault();
    const type = this.props.type;
    const formData = {
      email: this.state.formData.email,
      password: this.state.formData.password
    };
    if (type === "Register") {
      formData.username = this.state.formData.username;
    }
    const url = `${process.env.REACT_APP_USERS_SERVICE_URL}/auth/${type.toLowerCase()}`;
    axios
      .post(url, formData)
      .then((res) => {
        this.clearForm();
        this.props.loginUser(res.data.auth_token);
      })
      .catch((err) => {
        if (type === "Login") {
          this.props.createMessage("User does not exist.", "danger");
        }
        if (type === "Register") {
          this.props.createMessage("That user already exists.", "danger");
        }
      });
  }
  validateForm() {
    const formData = this.state.formData;
    this.resetRules();
    if (this.props.type === "Register") {
      const rules = this.state.registerFormRules;
      if (formData.username.length > 5) rules[0].valid = true;
      if (formData.email.length > 5) rules[1].valid = true;
      if (this.validateEmail(formData.email)) rules[2].valid = true;
      if (formData.password.length > 10) rules[3].valid = true;
      this.setState({ registerFormRules: rules });
      if (this.allValid()) this.setState({ valid: true });
    }
    if (this.props.type === "Login") {
      const rules = this.state.loginFormRules;
      if (formData.email.length > 0) rules[0].valid = true;
      if (formData.password.length > 0) rules[1].valid = true;
      this.setState({ loginFormRules: rules });
      if (this.allValid()) this.setState({ valid: true });
    }
  }
  allValid() {
    let rules = loginFormRules;
    if (this.props.type === "Register") {
      rules = registerFormRules;
    }
    for (let rule of rules) {
      if (!rule.valid) return false;
    }
    return true;
  }
  resetRules() {
    const registerFormRules = this.state.registerFormRules;
    for (let rule of registerFormRules) {
      rule.valid = false;
    }
    this.setState({ registerFormRules: registerFormRules });
    const loginFormRules = this.state.loginFormRules;
    for (let rule of loginFormRules) {
      rule.valid = false;
    }
    this.setState({ loginFormRules: loginFormRules });
    this.setState({ valid: false });
  }
  validateEmail(email) {
    // eslint-disable-next-line
    var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return re.test(email);
  }
  render() {
    if (this.props.isAuthenticated) {
      return <Redirect to="/" />;
    }
    let formRules = this.state.loginFormRules;
    if (this.props.type === "Register") {
      formRules = this.state.registerFormRules;
    }
    return (
      <>
        {this.props.type === "Login" && <h1 className="title">Log In</h1>}
        {this.props.type === "Register" && <h1 className="title">Register</h1>}
        <FormErrors type={this.props.type} rules={formRules} />
        <form onSubmit={this.handleSubmit}>
          {this.props.type === "Register" && (
            <div className="field">
              <label className="label">Username</label>
              <div className="control">
                <input
                  name="username"
                  value={this.state.username}
                  className="input"
                  type="text"
                  onChange={this.handleChange}
                />
              </div>
            </div>
          )}
          <div className="field">
            <label className="label">Email</label>
            <div className="control">
              <input
                name="email"
                value={this.state.email}
                className="input"
                type="email"
                onChange={this.handleChange}
                required
              />
            </div>
          </div>
          <div className="field">
            <label className="label">Password</label>
            <div className="control">
              <input
                name="password"
                value={this.state.password}
                className="input"
                type="password"
                onChange={this.handleChange}
                required
              />
            </div>
          </div>
          <button className="button" type="submit" disabled={!this.state.disabled}>
            Submit
          </button>
        </form>
      </>
    );
  }
}

export default Form;
