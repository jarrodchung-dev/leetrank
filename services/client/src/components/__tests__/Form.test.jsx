import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import Form from "../Form.jsx";

const forms = [
  {
    type: "Register",
    title: "Register",
    data: {
      username: "",
      email: "",
      password: ""
    },
    handleChange: jest.fn(),
    handleSubmit: jest.fn()
  },
  {
    type: "Login",
    title: "Log In",
    data: {
      email: "",
      password: ""
    },
    handleChange: jest.fn(),
    handleSubmit: jest.fn()
  }
];

forms.forEach((form) => {
  describe(`<${form.type} Form /> (not authenticated)`, () => {
    const component = (
      <Form
        type={form.type}
        data={form.data}
        handleChange={form.handleChange}
        handleSubmit={form.handleSubmit}
        isAuthenticated={false}
      />
    );
    const wrapper = shallow(component);

    it("renders the title of the form correctly", () => {
      const title = wrapper.find("h1.title");
      expect(title.length).toBe(1);
      expect(title.get(0).props.children).toBe(form.title);
    });

    it("renders only the fields that were passed in as props", () => {
      const fields = wrapper.find(".field");
      expect(fields.length).toBe(Object.keys(form.data).length);
      const prop = fields.get(0).props.children[1].props.children.props;
      expect(prop.name).toBe(Object.keys(form.data)[0]);
      expect(prop.value).toBe(Object.values(form.data)[0]);
    });

    it("submits submits the form correctly", () => {
      const input = wrapper.find(`input[type="email"]`);
      expect(form.handleChange).toHaveBeenCalledTimes(0);
      expect(form.handleSubmit).toHaveBeenCalledTimes(0);
      input.simulate("change");
      expect(form.handleChange).toHaveBeenCalledTimes(1);
      wrapper.find("form").simulate("submit", form.data);
      expect(form.handleSubmit).toHaveBeenCalledWith(form.data);
      expect(form.handleSubmit).toHaveBeenCalledTimes(1);
    });

    it("renders a snapshot successfully", () => {
      const tree = renderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe(`<${form.type} Form /> (authenticated)`, () => {
    const component = (
      <Form
        type={form.type}
        data={form.data}
        handleChange={form.handleChange}
        handleSubmit={form.handleSubmit}
        isAuthenticated={true}
      />
    );
    const wrapper = shallow(component);

    it("redirects away from <Form />", () => {
      expect(wrapper.find("Redirect")).toHaveLength(1);
    });
  });
});

/*


*/
