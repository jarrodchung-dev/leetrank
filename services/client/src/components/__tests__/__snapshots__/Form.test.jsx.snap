// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form renders a snapshot successfully 1`] = `
<div>
  <h1
    className="title"
  >
    Log In
  </h1>
  <form
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <label
        className="label"
      >
        Email
      </label>
      <div
        className="control"
      >
        <input
          className="input"
          name="email"
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Password
      </label>
      <div
        className="control"
      >
        <input
          className="input"
          name="password"
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      className="button"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Register Form renders a snapshot successfully 1`] = `
<div>
  <h1
    className="title"
  >
    Register
  </h1>
  <form
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <label
        className="label"
      >
        Username
      </label>
      <div
        className="control"
      >
        <input
          className="input"
          name="username"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Email
      </label>
      <div
        className="control"
      >
        <input
          className="input"
          name="email"
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Password
      </label>
      <div
        className="control"
      >
        <input
          className="input"
          name="password"
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      className="button"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
